### PR TITLE
Always allow nullable keys for projections

### LINKED ISSUE #54814: Create projection for MergeTree
**Describe the issue**
Creating projections for a MergeTree table was working well in the version `v23.3.8.21-lts` but now it is returning the error 

```
Sorting key contains nullable columns, but merge tree setting `allow_nullable_key` is disabled
```

**How to reproduce**
```sql
CREATE TABLE sales (
	DATE_SOLD DateTime64(3, 'UTC'),
	PRODUCT_ID Nullable(String),
) 
Engine MergeTree()
PARTITION BY toYYYYMM(DATE_SOLD)
ORDER BY DATE_SOLD
```

then create the projection
```sql
ALTER TABLE sales
    ADD PROJECTION test
    (
        SELECT toInt64(COUNT(*))
        GROUP BY
            PRODUCT_ID,
            DATE_SOLD
    )
```
It should create the projection, but instead is returning an error

* Which ClickHouse server versions are incompatible
Working versions: `23.3.8.21` and `23.5.1`
Broken in the version `23.8.2.7`

**Error message and/or stacktrace**
```
Received exception from server (version 23.8.2):
Code: 44. DB::Exception: Received from localhost:9000. DB::Exception: Sorting key contains nullable columns, but merge tree setting `allow_nullable_key` is disabled. (ILLEGAL_COLUMN)
```

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Always allow the creation of a projection with `Nullable` PK. This fixes https://github.com/ClickHouse/ClickHouse/issues/54814

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 973cd5e972192a2139c20191bf26c75c8a03183f with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54895/973cd5e972192a2139c20191bf26c75c8a03183f/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

File: tests/queries/0_stateless/01710_projection_with_nullable_keys.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+DROP TABLE IF EXISTS sales;
+
+CREATE TABLE sales (DATE_SOLD DateTime64(3, 'UTC'), PRODUCT_ID Nullable(String)) Engine MergeTree() PARTITION BY toYYYYMM(DATE_SOLD) ORDER BY DATE_SOLD;
+
+ALTER TABLE sales ADD PROJECTION test (SELECT toInt64(COUNT(*)) GROUP BY PRODUCT_ID, DATE_SOLD);
```
Comment (KochetovNicolai) [inline tests/queries/0_stateless/01710_projection_with_nullable_keys.sql:5]: I think this query should fail with `Sorting key contains nullable columns, but merge tree setting 'allow_nullable_key' is disabled`. This is expected when `allow_nullable_key` is not set.

But it should work after
```
alter table sales modify setting allow_nullable_key=1
```

Now it does not:
```
:) show create table sales

SHOW CREATE TABLE sales

Query id: 523f51e7-244c-4f8d-9445-b82e3718b43b

┌─statement───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CREATE TABLE default.sales
(
    `DATE_SOLD` DateTime64(3, 'UTC'),
    `PRODUCT_ID` Nullable(String)
)
ENGINE = MergeTree
PARTITION BY toYYYYMM(DATE_SOLD)
ORDER BY DATE_SOLD
SETTINGS index_granularity = 8192, allow_nullable_key = 1 │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

:) ALTER TABLE sales
    ADD PROJECTION test
    (
        SELECT toInt64(COUNT(*))
        GROUP BY
            PRODUCT_ID,
            DATE_SOLD
    )

ALTER TABLE sales
    ADD PROJECTION test
    (
        SELECT toInt64(COUNT(*))
        GROUP BY 
            PRODUCT_ID,
            DATE_SOLD
    )

Query id: 0101b5db-ccd3-4215-b70f-397d0b004d8d


0 rows in set. Elapsed: 0.002 sec. 

Received exception from server (version 23.9.1):
Code: 44. DB::Exception: Received from localhost:9000. DB::Exception: Sorting key contains nullable columns, but merge tree setting `allow_nullable_key` is disabled. (ILLEGAL_COLUMN)

```

And this ^ should be fixed.

Comment (KochetovNicolai) [reply tests/queries/0_stateless/01710_projection_with_nullable_keys.sql:5]: Btw, if you initially create a table with `allow_nullable_key = 1`, `ADD PROJECTION` works.

Comment (KochetovNicolai) [reply tests/queries/0_stateless/01710_projection_with_nullable_keys.sql:5]: Discussed it in TG.
The implementation looks ok if we look at projections as a kind of secondary index.
Let's make this change to make live easier.

**Review by KochetovNicolai — APPROVED** [review]

**Pull request merged**